### PR TITLE
Fix recurrence calculation

### PR DESCRIPTION
**Pull request by hubermat**
Solves issue 180.

Please run `composer test` to perform the unit tests.

### CODE REVIEW

**Comment by u01jmg3** [comment]
Hope to look at this soon but as there is quite a lot of new functionality it may take me some time.

**Comment by u01jmg3** [comment]
Ran your code and tests but I am getting 4 failures as well as `unserialize()` errors with PHPUnit itself:

```
There were 4 failures:

1) SingleEventsTest::testFullDayTimeZoneBerlin
Failed asserting that actual size 0 matches expected size 1.

\ics-parser\tests\SingleEventsTest.php:78
\ics-parser\tests\SingleEventsTest.php:18

2) SingleEventsTest::testSeveralFullDaysTimeZoneBerlin
Failed asserting that actual size 0 matches expected size 1.

\ics-parser\tests\SingleEventsTest.php:78
\ics-parser\tests\SingleEventsTest.php:32

3) SingleEventsTest::testEventTimeZoneUTC
Failed asserting that actual size 0 matches expected size 1.

\ics-parser\tests\SingleEventsTest.php:78
\ics-parser\tests\SingleEventsTest.php:45

4) SingleEventsTest::testEventTimeZoneBerlin
Failed asserting that actual size 0 matches expected size 1.

\ics-parser\tests\SingleEventsTest.php:78
\ics-parser\tests\SingleEventsTest.php:59

FAILURES!
Tests: 14, Assertions: 4, Errors: 10, Failures: 4.
Script phpunit --colors=always handling the test event returned with error code 2
```

File: src/ICal/ICal.php [hunk 1]
```diff
@@ -762,13 +762,7 @@ public function iCalDateToDateTime($icalDate, $forceTimeZone = false, $forceUtc
...
     {
         $dateTime = $this->iCalDateToDateTime($icalDate, $forceTimeZone, $forceUtc);
-        $offset   = 0;
-
-        if ($forceTimeZone) {
-            $offset = $dateTime->getOffset();
-        }
-
-        return $dateTime->getTimestamp() + $offset;
+        return $dateTime->getTimestamp();
```
Comment (u01jmg3) [inline src/ICal/ICal.php]: I will need some explanation on this change please

File: src/ICal/ICal.php [hunk 2]
```diff
@@ -1254,51 +1250,53 @@ protected function processRecurrences()
...
-                                    $anEvent['DTSTART_array'][1] = $anEvent['DTSTART'];
-                                    $anEvent['DTSTART_array'][2] = $monthRecurringTimestamp;
-                                    $anEvent['DTEND_array']      = $anEvent['DTSTART_array'];
-                                    $anEvent['DTEND_array'][2]  += $eventTimestampOffset;
-                                    $anEvent['DTEND'] = date(
-                                            self::DATE_TIME_FORMAT,
-                                            $anEvent['DTEND_array'][2]
-                                        ) . ($isAllDayEvent || ($initialEndTimeZoneName === 'Z') ? 'Z' : '');
-                                    $anEvent['DTEND_array'][1] = $anEvent['DTEND'];
+                                    if (($monthRecurringTimestamp > $startTimestamp) && ($monthRecurringTimestamp <= $until)) {
```
Comment (u01jmg3) [inline src/ICal/ICal.php]: Can you give some explanation on the requirement of this `if` condition please?

File: src/ICal/ICal.php [hunk 3]
```diff
@@ -1664,6 +1662,7 @@ protected function processEventIcalDateTime(array $event, $index = 3)
             if (isset($event["{$type}_array"])) {
                 $timeZone = (isset($event["{$type}_array"][0]['TZID'])) ? $event["{$type}_array"][0]['TZID'] : $calendarTimeZone;
                 $event["{$type}_array"][$index] = ((is_null($timeZone)) ? '' : sprintf(self::ICAL_DATE_TIME_TEMPLATE, $timeZone)) . $event["{$type}_array"][1];
+                $event["{$type}_array"][2] = $this->iCalDateToUnixTimestamp($event["{$type}_array"][3], true, true);
```
Comment (u01jmg3) [inline src/ICal/ICal.php]: Can you give some explanation on the requirement of this new key condition please?

**Comment by hubermat** [comment]
I am closing this pull request and creating two new ones, as this pull request is hard to bring in sync with current master (to be merged). 
New pull requests will be separate: introducing unit testing, and fixing the recurrence calculation so that the unit tests run without failures.

**Pull request closed without merging**